synapse/handlers/sync.py

```diff
@@ -1472,11 +1541,22 @@ def _calculate_state(timeline_contains, timeline_start, previous, current):
     }
 
     c_ids = set(e for e in current.values())
-    tc_ids = set(e for e in timeline_contains.values())
-    p_ids = set(e for e in previous.values())
     ts_ids = set(e for e in timeline_start.values())
+    p_ids = set(e for e in previous.values())
+    tc_ids = set(e for e in timeline_contains.values())
+
+    # track the membership events in the state as of the start of the timeline
+    # so we can add them back in to the state if we're lazyloading.  We don't
+    # add them into state if they're already contained in the timeline.
+    if lazy_load_members:
+        ll_ids = set(
```

Comment:
I wonder if this would be clearer rewritten as:

```
p.ids -= set (e for t, e in timeline_start.iteritems() if t[0] == EventTypes.Member)
```

or, more efficiently, especially when p_ids is empty, as it is for the initialsync case:

```
p_ids.difference_update(e for t, e in timeline_start.iteritems() if t[0] == EventTypes.Member)
```

Reply:
i don't understand why you are trying to remove stuff from `p.ids`, which are the IDs of the state at the end of the previous sync (if any).  the idea here is to add state for the lazy-loaded members into the result, not remove stuff from the previous sync result, where it probably doesn't exist in the first place?

Reply:
aren't the two equivalent?

Reply:
ok.